wallet/src/signer/trezor_signer/mod.rs

```diff
@@ -1240,61 +1513,146 @@ fn check_public_keys(
             HardwareWalletData::Trezor(data) => {
                 // If the device_id is the same but public keys are different, maybe a
                 // different passphrase was used
-                if data.device_id == provider.data.device_id {
-                    return Err(WalletError::HardwareWalletDifferentPassphrase);
+                if data.device_id == trezor_data.device_id {
+                    return Err(TrezorError::HardwareWalletDifferentPassphrase.into());
                 } else {
-                    return Err(WalletError::HardwareWalletDifferentMnemonicOrPassphrase {
+                    return Err(TrezorError::HardwareWalletDifferentMnemonicOrPassphrase {
                         file_device_id: data.device_id,
-                        connected_device_id: provider.data.device_id.clone(),
+                        connected_device_id: trezor_data.device_id.clone(),
                         file_label: data.label,
-                        connected_device_label: provider.data.label.clone(),
-                    });
+                        connected_device_label: trezor_data.label.clone(),
+                    }
+                    .into());
                 }
             }
         }
     }
 
-    Err(WalletError::HardwareWalletDifferentFile)
+    Err(TrezorError::HardwareWalletDifferentFile)?
 }
 
-fn find_trezor_device() -> Result<(Trezor, TrezorData), TrezorError> {
-    let mut devices = find_devices(false)
+fn fetch_extended_pub_key(
+    client: &mut Trezor,
+    chain_config: &ChainConfig,
+    account_index: U31,
+) -> Result<ExtendedPublicKey, TrezorError> {
+    let derivation_path = make_account_path(chain_config, account_index);
+    let account_path = derivation_path.as_slice().iter().map(|c| c.into_encoded_index()).collect();
+    let chain_type = to_trezor_chain_type(chain_config);
+    let xpub = client
+        .mintlayer_get_public_key(chain_type, account_path)
+        .map_err(|e| TrezorError::DeviceError(e.to_string()))?;
+    let chain_code = ChainCode::from(xpub.chain_code.0);
+    let account_pubkey = Secp256k1ExtendedPublicKey::new_unchecked(
+        derivation_path,
+        chain_code,
+        Secp256k1PublicKey::from_bytes(&xpub.public_key.serialize())
+            .map_err(|_| TrezorError::InvalidKey)?,
+    );
+    let account_pubkey = ExtendedPublicKey::new(account_pubkey);
+    Ok(account_pubkey)
+}
+
+/// Check that the public keys in the DB are the same as the ones with the connected hardware
```

Comment:
Probably "with" should be "from", same for `check_public_keys_against_key_chain`.